keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+- A single custom resource is defined with the list of banned words (but has not
+  matching rules of its own).
+
+
+```yaml
+apiVersion: rules.example.com/v1
+kind: BannedWords
+metadata:
+  name: "banned-words.example.com"
+spec:
+  bannedWords:
+  - glitter
+  - rainbow
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.name in config.bannedWords"
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy2.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.spec.containers.any(c, c.name in config.bannedNames)"
+  - rule: "!object.spec.initContainers.any(c, c.name in config.bannedNames)"
+```
+
+
+Similar Use Case: A cluster administrator wishes to use a single policy
+configuration to manage a network policy that must be enforced across multiple
+Kubernetes kinds that contain relevant networking fields. It is possible to
+implement by having multiple `ValidatingAdmissionPolicy` resources that all
+reference the same `spec.config` CRD but that each enforce the policy for a
+different Kubernetes network kind.
+
+#### Use Case: Principle of least privilege policy
+
+A cluster administrator would like disallow the use of a list of reserved labels
+by default, but allow use of the labels in specific namespaces so long as the
+label values are valid. 
+
+- Define the following policy configurations:
+  - policy configuration that matches all namespaces with the
+    "may-use-reserved-labels" label and checks that the values are valid
+  - "default" policy configuration that matches all namespaces without the
+    "may-use-reserved-labels" label and disallows any reserved labels.
+  - An additional policy can be used validate that only authorized roles set the
+    "may-use-reserved-labels".
+
+See "Default configuration
+
+#### Use Case: Validating native type with new field (version skew case)
+
+Policy author wants to write a policy that validates a property of all
+containers in a pod, including ephemeralContainers for versions of
+Kubernetes where ephemeralContainers are available.
+
+```yaml
+  validations:
+    - expression: "object.spec.containers.all(c, c.name.startsWith('xyz-'))"
+    - expression: "!has(object.spec.initContainers) || object.spec.initContainers.all(c, c.name.startsWith('xyz-'))"
+    - expression: "!has(object.spec.ephemeralContainers) || object.spec.ephemeralContainers.all(c, c.name.startsWith('xyz-'))"
+```
+
+This does not work if type checking is strict. The ephemeralContainers field
```

Comment:
`!has(object.spec.ephemeralContainers)` doesn't function properly with strict type checking?  Is there a "does this field exist in the schema" function we can use during execution?

Reply:
Good question.

Unfortunately no, CEL doesn't offer a way to do that. Type checking is optional and if run, it produces any errors it finds, but it does not supplement the "compiled" CEL expression with additional type information to use at runtime.

Reply:
Is there any way to expose the `ephemeralContainers` field as a dynamically typed object that's exposed by default on the spec? Effectively extend the type provided with one that layers in some added metadata?